parser/src/json/numeric.rs

```diff
@@ -0,0 +1,364 @@
+                x.chars().next().unwrap(),
+                lexi_x_to_9(x_rest, incl)
+            )];
+            if x0 < 9 {
+                parts.push(format!("[{}-9][0-9]*", x0 + 1));
+            }
+            mk_or(parts)
+        }
+    } else {
+        if x.is_empty() {
+            "[0-9]*[1-9]".to_string()
+        } else {
+            let x0 = x.chars().next().unwrap().to_digit(10).unwrap();
+            let x_rest = &x[1..];
+            let mut parts = vec![format!(
+                "{}{}",
+                x.chars().next().unwrap(),
+                lexi_x_to_9(x_rest, incl)
+            )];
+            if x0 < 9 {
+                parts.push(format!("[{}-9][0-9]*", x0 + 1));
+            }
+            mk_or(parts)
+        }
+    }
+}
+
+fn lexi_0_to_x(x: &str, incl: bool) -> String {
+    if x.is_empty() {
+        assert!(incl);
+        "".to_string()
+    } else {
+        let x0 = x.chars().next().unwrap().to_digit(10).unwrap();
+        let x_rest = &x[1..];
+
+        if !incl && x.len() == 1 {
+            assert!(x0 > 0);
+            return format!("[0-{}][0-9]*", x0 - 1);
+        }
+
+        let mut parts = vec![format!(
+            "{}{}",
+            x.chars().next().unwrap(),
+            lexi_0_to_x(x_rest, incl)
+        )];
+        if x0 > 0 {
+            parts.push(format!("[0-{}][0-9]*", x0 - 1));
+        }
+        mk_or(parts)
+    }
+}
+
+fn lexi_range(ld: &str, rd: &str, ld_incl: bool, rd_incl: bool) -> String {
+    assert_eq!(ld.len(), rd.len());
+    if ld == rd {
+        assert!(ld_incl && rd_incl);
+        ld.to_string()
+    } else {
+        let l0 = ld.chars().next().unwrap().to_digit(10).unwrap();
+        let r0 = rd.chars().next().unwrap().to_digit(10).unwrap();
+        if l0 == r0 {
+            let ld_rest = &ld[1..];
+            let rd_rest = &rd[1..];
+            format!(
+                "{}{}",
+                ld.chars().next().unwrap(),
+                lexi_range(ld_rest, rd_rest, ld_incl, rd_incl)
+            )
+        } else {
+            assert!(l0 < r0);
+            let ld_rest = ld[1..].trim_end_matches('0');
+            let mut parts = vec![format!(
+                "{}{}",
+                ld.chars().next().unwrap(),
+                lexi_x_to_9(ld_rest, ld_incl)
+            )];
+            if l0 + 1 < r0 {
+                parts.push(format!("[{}-{}][0-9]*", l0 + 1, r0 - 1));
+            }
+            let rd_rest = rd[1..].trim_end_matches('0');
+            if !rd_rest.is_empty() || rd_incl {
+                parts.push(format!(
+                    "{}{}",
+                    rd.chars().next().unwrap(),
+                    lexi_0_to_x(rd_rest, rd_incl)
+                ));
+            }
+            mk_or(parts)
+        }
+    }
+}
+
+fn float_to_str(f: f64) -> String {
+    format!("{}", f)
+}
+
+pub fn rx_float_range(
+    left: Option<f64>,
+    right: Option<f64>,
+    left_inclusive: bool,
+    right_inclusive: bool,
+) -> String {
```

Comment:
Should be easy enough to use `anyhow::Result<String>` and than `ensure!()` instead of `assert!()` and `bail!()` instead of `panic!()`, at least in the top-level function. The inner asserts should probably never happen.